### DISCUSSION TITLE
Cannot reproduce LinkNeighborSampling

### DISCUSSION THREAD

**Discussion by kei0822kei** [pyg-team/pytorch_geometric#9878] [Q&A]
# 🐛 Describe the bug

Hi, I am struggling to reproduce batch sampling with LinkNeighborLoader.

Following is what I tried:
```python
import torch_geometric.transforms as T
from torch_geometric.datasets import OGB_MAG
from torch_geometric.loader import NeighborLoader
from torch_geometric import seed_everything


def build_batch(seed):
    seed_everything(seed)
    transform = T.ToUndirected()  # Add reverse edge types.
    data = OGB_MAG(root="./data", preprocess="metapath2vec", transform=transform)[0]
    train_loader = NeighborLoader(
        data,
        # Sample 15 neighbors for each node and each edge type for 2 iterations:
        num_neighbors=[15] * 2,
        # Use a batch size of 128 for sampling training nodes of type "paper":
        batch_size=128,
        input_nodes=("paper", data["paper"].train_mask),
    )
    batch = next(iter(train_loader))
    return batch


batch1 = build_batch(42)
batch2 = build_batch(42)
print(batch1["field_of_study", "rev_has_topic", "paper"].edge_index.shape)
print(batch2["field_of_study", "rev_has_topic", "paper"].edge_index.shape)
```

```text:output
torch.Size([2, 10472])
torch.Size([2, 10458])
```

I expected `batch1` and `batch2` are identical, but they are not.

How do I fix this problem?

# Environment
```text
torch                              2.3.0+cu121
torch_geometric                    2.5.3
```

**Comment by Romain-Nicolle** [comment]
Hi,

I am not an expert but I will try to help.
Someone had already posted a question about the reproductibility of the code using the random seed, in the general discussion, topic 8716 [here](https://github.com/pyg-team/pytorch_geometric/discussions/8716).
Basically they rewrote the function to seed everything to include more parameters to set the CUDA behavior to deterministic.

The complete function is here:
```
import math
import os
import random
import time
import numpy as np
import torch

def set_random_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)  
    torch.cuda.manual_seed(seed)  
    torch.cuda.manual_seed_all(seed) 
    os.environ['PYTHONHASHSEED'] = str(seed)
    torch.backends.cudnn.deterministic = True  
    torch.backends.cudnn.benchmark = False
    torch.backends.cudnn.enabled = True
```

There were more comments on this post, so I advice you go there to check.
But I hope this should help you overcome your problem.

best,

**Comment by kei0822kei** [comment]
@Romain-Nicolle 

Thanks a lot for replying and telling me a helpful post.

I revised my code as:

```python
import os
import random
import numpy as np
import torch
import torch_geometric.transforms as T
from torch_geometric.datasets import OGB_MAG
from torch_geometric.loader import NeighborLoader
from torch_geometric import seed_everything


def set_random_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    os.environ["PYTHONHASHSEED"] = str(seed)
    torch.backends.cudnn.enabled = True
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
    torch.use_deterministic_algorithms(True)


def build_batch(seed):
    set_random_seed(seed)
    transform = T.ToUndirected()  # Add reverse edge types.
    data = OGB_MAG(root="./data", preprocess="metapath2vec", transform=transform)[0]
    train_loader = NeighborLoader(
        data,
        # Sample 15 neighbors for each node and each edge type for 2 iterations:
        num_neighbors=[15] * 2,
        # Use a batch size of 128 for sampling training nodes of type "paper":
        batch_size=128,
        input_nodes=("paper", data["paper"].train_mask),
    )
    batch = next(iter(train_loader))
    return batch


batch1 = build_batch(42)
batch2 = build_batch(42)
print(batch1["field_of_study", "rev_has_topic", "paper"].edge_index.shape)
print(batch2["field_of_study", "rev_has_topic", "paper"].edge_index.shape)
```

but output keeps non-reproducible.

```text
torch.Size([2, 10466])
torch.Size([2, 10472])
```

Maybe I should check deeper, but your comment was so helpful. Thanks a lot.
